changes/2024-TBD_cache-across-hierarchy-keyrings/change.md

```diff
@@ -0,0 +1,552 @@
+to the Hierarchy Keyring at construction,
+the CMC MAY cache Branch Keys protected by different
+KMS Relationships.
+
+This simplifies Multiple KMS Relationship MPL Consumers,
+as they do not need to stand up a LRU Cache of Hierarchy Keyrings.
+
+Instead, they may maintain one CMC.
+They still create a Hierarchy Keyring instance per KMS Relationship,
+and they MUST use the correct Keyring to retrieve material
+from the Cache.
+
+But they only need to maintain the common cache.
+
+In the future, the CachingCMM will be introduced to
+Crypto Tool's Dafny products.
+The CachingCMM and the Hierarchy Keyring both consume CMCs.
+Thus, it will be possible to provide a CMC to both a
+Hierarchy Keyring & a CachingCMM.
+This cache identifier design MUST consider
+how the cache entries of multiple Hierarchy Keyrings
+and CachingCMMs will be appropriately separated.
+
+## Requirements
+
+- We must allow caching across Key Stores/KMS Clients/KMS Keys
+  for multiple Hierarchy Keyrings and CachingCMMs.
+  For this, we need to strategically update the cache identifiers
+  for all materials ([background](#background)) in the CMC.
+  The [Issues and Alternatives section](#issues-and-alternatives)
+  discusses this.
+
+## Out of Scope
+
+- Supporting a shared Cache in the DB-ESDK to fetch
+  BeaconKeyMaterials across multiple Searchable
+  Encryption Configurations, or across multiple KMS Relationships.
+
+## Issues and Alternatives:
+
+Before looking at the issues, please take a quick look at
+[Hierarchy Keyring Cache Identifier formula](#hierarchy-keyring-cache-identifier-formula),
+[CachingCMM Cache Identifier formula](#cachingcmm-cache-identifier-formula), and
+[DB-ESDK Searchable Encryption Cache Identifier formula](#db-esdk-searchable-encryption-cache-identifier-formula) for more context.
+
+Note that there are two hashings taking place.
+For example in the Hierarchy Keyring Cache Identifier formula,
+one hash is for the branch-key-id and the other is hashing
+the entire appended string
+(a part of which is the hash of the branch-key-id)
+to get the final cache identifier.
+We'll call these internal and external hashes.
+
+#### Issue 1: How to update the cache identifier for the Hierarchy Keyring to allow a shared cache?
+
+Note that according to the current
+[Hierarchy Keyring Cache Identifier formula](#hierarchy-keyring-cache-identifier-formula),
+if two KMS Relationships Hierarchy Keyrings are
+called with two different Key Stores but the same branch-key-id,
+there will be a collision (the branch-key-version will also
+have to co-incide in case of Decryption Branch Key Materials).
+
+##### Option 1 (Recommended): Append the keystore-id to the [Hierarchy Keyring cache identifier formula](#hierarchy-keyring-cache-identifier-formula).
+
+By appending the keystore-id to the
+Hierarchy Keyring cache identifier,
+we can ensure that there are no collisions between two identifiers
+from different Key Stores.
```

Comment:
Okay, sounds good. Thanks.

Adding a slightly modified version of this to the doc based on our slack conversation
'By adding Key Store ID, we let the customer control collisions.
I.e: they choose if different KMS Relationships CAN hit the same identifier or not.
They can also choose if different logical key stores or physical key stores can collide or not.
But by default, because each Key Store object will (likely) have a unique UUID,
there will be no collision.'